napari/view_layers.py

```diff
@@ -1,165 +1,172 @@
-                add_string=add_string,
-            )
-        )
-
-    # get signatures of the add_* method and Viewer.__init__
-    add_sig = inspect.signature(add_method)
-    view_sig = inspect.signature(Viewer)
-
-    # create a new combined signature
-    new_params = list(add_sig.parameters.values())[1:]  # [1:] to remove self
-    new_params += view_sig.parameters.values()
-    new_params = sorted(new_params, key=lambda p: p.kind)
-    combo_sig = add_sig.replace(parameters=new_params)
-
-    # make new function string with the combined signature
-    fakefunc = f"def view_{layer_string}{combo_sig}:"
-    fakefunc += """
-        _kw = locals()
-        view_kwargs = {
-            k: _kw.pop(k) for k in list(_kw) if k in view_sig.parameters
-        }
-        viewer = napari.Viewer(**view_kwargs)
-        if 'kwargs' in _kw:
-            _kw.update(_kw.pop("kwargs"))
+    func : callable
+        The same function, with merged metadata.
     """
-    fakefunc += f"    viewer.{add_string}(**_kw)\n        return viewer"
-    # evaluate the new function into the current module namespace
-    globals = sys.modules[__name__].__dict__
-    eval(
-        compile(fakefunc, __file__, "exec"),
-        {
-            'typing': typing,
-            'view_sig': view_sig,
-            'Union': typing.Union,
-            'Optional': typing.Optional,
-            'List': typing.List,
-            'NoneType': type(None),
-            'Sequence': typing.Sequence,
-            'napari': sys.modules.get('napari'),
-        },
-        globals,
+    from .utils.misc import combine_signatures
+
+    # get the `Viewer.add_*` method
+    layer_string = func.__name__.replace("view_", "")
+    if layer_string == 'path':
+        add_method = Viewer.open
+    else:
+        add_method = getattr(Viewer, f'add_{layer_string}')
+
+    # merge the docstrings of Viewer and viewer.add_*
+    func.__doc__ = _merge_docstrings(add_method, layer_string)
+
+    # merge the signatures of Viewer and viewer.add_*
+    func.__signature__ = combine_signatures(
+        add_method, Viewer, return_annotation=Viewer, exclude=('self',)
     )
-    view_func = globals[f'view_{layer_string}']  # this is the final function.
-    view_func.__doc__ = merge_docs(add_method, layer_string)
-
-
-for _layer in (
-    'image',
-    'points',
-    'labels',
-    'shapes',
-    'surface',
-    'vectors',
-    'tracks',
-):
-    _generate_view_function(_layer)
-
-_generate_view_function('path', 'open')
+
+    # merge the __annotations__
+    func.__annotations__ = {
+        **add_method.__annotations__,
+        **Viewer.__init__.__annotations__,
+        'return': Viewer,
+    }
+
+    # update function's __globals__ ... Careful!
+    # this actually updates the globals for the entire view_layers module
+    # (but this is important for evaluating ForwardRef type hints later)
+    func.__globals__.update(
+        {**add_method.__globals__, **Viewer.__init__.__globals__, **globals()}
+    )
```

Comment:
refactored.  The reasoning for this was that when you call `typing.get_type_hints(obj)` (as is done in stubgen.py), it uses `obj.__globals__` to resolve any forward references.  So, this was an attempt to "attach" the necessary namespace to the function object so that get_type_hints can resolve them later.
I updated it to use a special attribute (`func._forwardrefns_` instead of `func.__globals__`) and modified `stubgen.py` to look for it.